Allow serializing and passing of credentials_hashes in DeviceConfig

sdb9696 opened the pull request:
As per [HA PR discussion](https://github.com/home-assistant/core/pull/105143#discussion_r1439889873) this PR enables the retrieval of the credentials hash and passing back into the `DeviceConfig`.  This will allow devices to persist all of their config without storing clear text credentials for later retrieval and re-use.

rytilahti commented on kasa/deviceconfig.py:
```diff
@@ -117,12 +129,11 @@ class DeviceConfig:
-    credentials: Credentials = field(
-        default_factory=lambda: Credentials(username="", password="")
-    )
+    credentials: Optional[Credentials] = None
+    credentials_hash: Optional[str] = None
```
We should start slowly documenting these (using the format like shown in https://github.com/python-kasa/python-kasa/blob/master/kasa/smartbulb.py#L94-L97 which produces https://python-kasa.readthedocs.io/en/stable/smartbulb.html#kasa.TurnOnBehaviors), but that can wait for another PR, just something to keep in mind wrt #600.

rytilahti commented on kasa/deviceconfig.py:
```diff
@@ -131,15 +142,22 @@ class DeviceConfig:
+    def to_dict(
+        self,
+        *,
+        credentials_hash: Optional[str] = None,
+        exclude_credentials: bool = False,
```
Maybe it'd be better to always export only the credentials hash, and get rid off these arguments?

sdb9696 replied:
It's needed to provide the hash as a parameter because the hash is in the transport itself and not the device config.  I think the parameters are good to have because they give the options of credentials, no credentials and hashes which covers all use cases.

rytilahti replied:
Hmmm, I think the question is should the transport details then also be stored in the exported config? It's again about trying to keep the API as simple as straightfoward as possible for downstreams, now it's sort of unclear why some parts are stored inside the config and some are passed in when exporting etc...

rytilahti reviewed (commented):
Looks fairly straightforward, some commentary inline.

sdb9696 commented on kasa/klaptransport.py:
```diff
@@ -119,6 +125,11 @@ def default_port(self):
         return self.DEFAULT_PORT
 
+    @property
+    def credentials_hash(self) -> Optional[str]:
```
Done

sdb9696 commented on kasa/aestransport.py:
```diff
@@ -80,6 +91,11 @@ def default_port(self):
         return self.DEFAULT_PORT
 
+    @property
+    def credentials_hash(self) -> Optional[str]:
```
Done

sdb9696 commented on kasa/aestransport.py:
```diff
@@ -88,6 +104,12 @@ def _http_client(self) -> httpx.AsyncClient:
             self._default_http_client = httpx.AsyncClient()
         return self._default_http_client
 
+    def get_login_params(self):
```
Done

sdb9696 commented on kasa/cli.py:
```diff
@@ -214,6 +220,12 @@ def _device_to_serializable(val: SmartDevice):
+@click.option(
+    "--credentials-hash",
+    default=None,
+    required=False,
```
Done

sdb9696 commented on kasa/cli.py:
```diff
@@ -184,6 +184,12 @@ def _device_to_serializable(val: SmartDevice):
+@click.option(
+    "--login-version",
+    envvar="KASA_LOGIN_VERSION",
+    default=None,
```
I've kept this as defaulting to `None` as it then matches with the discovery response for pre `lv:2` which doesn't have the `lv` at all.  This makes it cleaner elsewhere as I tried defaulting to 1 at the beginning but it got messy.  However I've made the `login_version` optional in the cli so just the encrypt type and device family is needed to control devices in cases where the version is irrelevant, so hopefully this is addressed.

rytilahti replied:
Okay, let's go with that!

sdb9696 commented on kasa/protocol.py:
```diff
@@ -56,13 +56,19 @@ def __init__(
 
+    @property
+    @abstractmethod
+    def credentials_hash(self) -> Optional[str]:
```
Done

sdb9696 commented on kasa/protocol.py:
```diff
@@ -120,6 +126,11 @@ def default_port(self):
+    @property
+    def credentials_hash(self) -> Optional[str]:
+        """The hashed credentials used by the transport."""
+        return None
```
This was already abstract but now the `_XorTransport` returns empty string so the Optional signature is completely dropped.

sdb9696 commented:
Hi, addressed most comments except one which I hope will be ok with you.

Great to see https://github.com/python-kasa/python-kasa/pull/594 going in.  Despite the challenges for @gimpy88 over testing there ended up being quite a small amount of code changes to get it working which is reassuring!

@rytilahti I think once this is merged we could cut 0.6.0 and then we're ready to move onto finalising the HA PR.  Hope you agree :-)

rytilahti commented:
Sounds good to me! :+1: I'm going to be occupied for the next couple of weeks, so it may be that I'm not able to help much before the second half of the month. The issues marked to the milestone are what I think makes sense to get in before cutting the release, one potential addition being some groundwork for the modules/components support as that'd make contributing individual features much easier by avoiding changes all across the tapodevice class.

sdb9696 commented:
Ok cool, I will start picking up those issues. How would you feel about cutting an alpha release before you go so I can start getting the HA release passing CI and coverage checks etc. At the moment it fails checks right at the start.

rytilahti commented:
I'm wondering if the CI would allow using a git reference to a specific commit? But sure, I could cut an alpha release. Would you like to have this PR before I do that?

sdb9696 commented:
Yes please.

rytilahti commented:
Release now available on pypi https://pypi.org/project/python-kasa/0.6.0.dev0/ :tada:

sdb9696 commented:
> Sounds good to me! 👍 I am also feeling positive that the effort we put earlier starts bearing fruits given how easily adding the support for those wall switches turned out to be in the end :-) I just ordered a L900 led strip as it was pretty cheap, I'm hoping to get that also in so that we'd have all but powerstrips & hub-based devices covered for the initial release.

Thanks for the alpha release @rytilahti.  N.B. the HA PR is updated to point to the alpha and my son's L900 that arrived yesterday worked straight away as it reports as `SMART.TAPOBULB` (for switching on and off at least)

> I'm going to be occupied for the next couple of weeks, so it may be that I'm not able to help much before the second half of the month. The issues marked to the milestone are what I think makes sense to get in before cutting the release, one potential addition being some groundwork for the modules/components support as that'd make contributing individual features much easier by avoiding changes all across the tapodevice class.

https://github.com/python-kasa/python-kasa/pull/614 is created for the docs.  I'll start on the test coverage next.

My preference would be to do the feature groundwork after 0.6.0 as that may take some time and back and forth.  Especially if you're unavailable for the next couple of weeks which could delay us getting the HA PR in for the Feb release.

rytilahti commented:
@sdb9696 sure thing! I received my L900 today and wanted to play with it a bit. Looks like the account "test@tp-link.net" using password "test" works for querying the device, and I was even able to provision it. Changing the password using a query similar to `wifi join` (passing the "account" and "time" for `set_qs_info`) worked _once_ even when it errored out with a json decode error, but I'm unable to change it again.

> My preference would be to do the feature groundwork after 0.6.0 as that may take some time and back and forth. Especially if you're unavailable for the next couple of weeks which could delay us getting the HA PR in for the Feb release.

Makes sense, adding modules won't likely introduce any immediate breaking changes so we can do that in the 0.6 series :-)
